IIgs SCC: Allow choosing 115200bps as the card allows it

colinleroy opened the pull request:
The same, for IIgs.

@oliverschmidt : I didn't add IRQ-disabling code. This should work all the way up to 115k accelerated at 8MHz or more. But I can add it if necessary.

colinleroy commented on libsrc/apple2/ser/a2.gs.s:
```diff
@@ -127,7 +138,7 @@ BaudTable:                              ; bit7 = 1 means setting is invalid
                 .byte   $06             ; SER_BAUD_38400
                 .byte   $07             ; SER_BAUD_57600
-                .byte   $FF             ; SER_BAUD_115200
+                .byte   $00             ; SER_BAUD_115200
```
It does not in this way - the baud rate generator's time constants are not even loaded in 115200 mode. Instead, another register instructs the chip to use the /RTxC and /SYNC pins to drive the clock. 
Maybe that warrants a comment at this place?

oliverschmidt replied:
Looking at the bytes spent for getClockSource, I wondering if it might result in less code if here `$40` would be used and down below...
```
        tay
        lda     BaudTable,y             ; Get chip value from Low/High tables
        bpl     BaudOK                  ; Verify baudrate is supported
```
...would be replaced with...
```
        tax
        bit     BaudTable,x             ; Get chip value from Low/High tables
        bpl     BaudOK                  ; Verify baudrate is supported
        bvc/bvs ????
```
We can use the 65C02 specific `bit <abs>,x` as we know we're on a IIgs.

colinleroy replied:
I went to bed and will check tomorrow, but I think X is held up by the channel

colinleroy replied:
Yep, X is held up with the channel value. I've tried various experiments (clock source in bit 6 of BaudTable, or an addition to BaudLow/BaudHighTable with clock source as high bit), the best I could find was -4 bytes at the price of convoluted, hard to understand code.

colinleroy replied:
I can save 5 bytes using the fact that SER_PARAMS::BAUDRATE = 0 though :)

oliverschmidt commented:
I understand that my perspective can be considered inconsistent, but I agree with you to keep the IRQ enabled. It's more of a gut feeling:
* Significant acceleration is way more common on the IIgs than on the 8-bit machines.
* In contrast to the SSC there are 38400 and 57600 so the question would be where to draw the line.

colinleroy commented:
Doesn't seem inconsistent to me, in fact it did not add that directly because I had a feeling you would think that - and I agree!

colinleroy commented:
Here you are, kind of a big cleanup :) I've replaced channel-dependant branches with indexed variables, simplified logic where I could (for example the two levels of Baud*Table indexes was not ideal, and also there's no need to skip setting WR12/13 even when using XTAL mode, they're just ignored). While I was at it I made the comments clearer too (in my opinion).
With that, the 115k feature "only" costs 22 bytes.

oliverschmidt commented:
> Here you are, kind of a big cleanup :) [...]

Great - and thanks for the explanation :grinning:

colinleroy commented:
Hi Oliver,
is there anything missing from me for this one?
Thanks!

oliverschmidt commented:
I asked you this question:

> May I ask why you removed the bra here?

colinleroy commented:
> I asked you this question:
> 
> > May I ask why you removed the bra here?

Ah ! I did well to ask, because I had totally not seen this question (and TBH, I still don't see it. Probably a github issue). I suppose you asked about the bra at line 415/432 that I replaced with a beq. It's simply that I prefer avoiding using non 6502 opcodes when it's as easy to use 6502 opcodes. I know this may not make much sense in a 65816 file, but that helps me not get too used to them :)

oliverschmidt commented:
I see your point. I personally think that in this case an `; Always` comment is desirable. But that's too much nitpicking even for me. So if you want to add the comment, do so - otherwise I'll merge as-is.

colinleroy commented:
> I see your point. I personally think that in this case an `; Always` comment is desirable. But that's too much nitpicking even for me. So if you want to add the comment, do so - otherwise I'll merge as-is.

I may as well put the bra back to make it clear without being overly verbose, then: done :)
TBH, I skipped over that `bra` and noticed I had a known Y right before and I may as well... but it's not a hard feeling !
Thanks :)

oliverschmidt commented:
It's - as always - very nice to discuss things with you :satisfied:

colinleroy commented:
I can say the same thing! Thanks 🙂
